File: impl/golang/internal/account/account.go
package account

import (
	"math/big"
	"sync"

	"github.com/neuron-sdk/neuron-go-sdk/internal/keylib"
)

// FR-013: Three valid account types (Parent, Child, Shared)
// AccountType discriminates between the three account kinds.
// The zero value (Unspecified) is invalid and will be rejected by validation.
type AccountType int

const (
	// Unspecified is the zero value and is always invalid.
	Unspecified AccountType = iota
	// Parent is the root identity account with DID and credit balance.
	Parent
	// Child is an agent/device account referencing a Parent.
	Child
	// Shared is a multisig threshold account with MultisigKey.
	Shared
)

// String returns the human-readable name of the AccountType.
func (t AccountType) String() string {
	switch t {
	case Parent:
		return "Parent"
	case Child:
		return "Child"
	case Shared:
		return "Shared"
	default:
		return "Unspecified"
	}
}

// IsValid returns true for Parent, Child, and Shared; false for Unspecified or unknown values.
func (t AccountType) IsValid() bool {
	return t == Parent || t == Child || t == Shared
}

// LedgerAccountId is an opaque identifier for a ledger account used as a fee payer.
type LedgerAccountId string

// FR-001: Parent account cryptographic identity
// FR-002: Child account with parent reference
// NeuronAccount is the primary entity representing an agent identity in the Neuron network.
// It is constructed via type-safe builders (NewParentAccountBuilder, NewChildAccountBuilder,
// NewSharedAccountBuilder) and validated via Validate().
//
// Fields are type-specific — see data-model.md for the MUST/MUST NOT matrix.
type NeuronAccount struct {
	accountType     AccountType
	publicKey       *keylib.NeuronPublicKey
	evmAddress      *keylib.EVMAddress
	peerID          *keylib.PeerID
	did             *NeuronDID
	parentPubKey    *keylib.NeuronPublicKey
	multisigKey     *keylib.MultisigKey
	currencySymbol  string
	creditBalance   *big.Int // Parent only; nil until ledger sync
	balanceAllocat  *big.Int // Child only; nil until ledger sync
	balance         *big.Int // Shared only; nil until ledger sync
	ledgerAttach    *LedgerAttachment
	registryBinding *RegistryBinding
	feePayer        *LedgerAccountId
	p2pHost         *keylib.PeerID // Child only; = peerID
	mu              sync.Mutex
}

// AccountType returns the type discriminator (Parent, Child, or Shared).
func (a *NeuronAccount) AccountType() AccountType {
	return a.accountType
}

// PublicKey returns the account's NeuronPublicKey, or nil for Shared accounts.
func (a *NeuronAccount) PublicKey() *keylib.NeuronPublicKey {
	return a.publicKey
}

// FR-008: Identity derivation from public key
// EVMAddress returns the derived EVM address, or nil for Shared accounts.
func (a *NeuronAccount) EVMAddress() *keylib.EVMAddress {
	return a.evmAddress
}

// PeerID returns the derived PeerID, or nil for Shared accounts.
func (a *NeuronAccount) PeerID() *keylib.PeerID {
	return a.peerID
}

// DID returns the NeuronDID, or nil for Child and Shared accounts.
func (a *NeuronAccount) DID() *NeuronDID {
	return a.did
}

// ParentPublicKey returns the parent's NeuronPublicKey for Child accounts, nil otherwise.
func (a *NeuronAccount) ParentPublicKey() *keylib.NeuronPublicKey {
	return a.parentPubKey
}

// FR-021: Shared accounts with MultisigKey
// MultisigKey returns the MultisigKey for Shared accounts, nil otherwise.
func (a *NeuronAccount) MultisigKey() *keylib.MultisigKey {
	return a.multisigKey
}

// FR-020: Currency symbol identification
// CurrencySymbol returns the currency symbol (e.g. "ETH", "HBAR").
func (a *NeuronAccount) CurrencySymbol() string {
	return a.currencySymbol
}

// FR-016: Balance management (credit, allocation)
// CreditBalance returns the Parent's credit balance, or nil if not yet synced.
func (a *NeuronAccount) CreditBalance() *big.Int {
	return a.creditBalance
}

// BalanceAllocation returns the Child's balance allocation, or nil if not yet synced.
func (a *NeuronAccount) BalanceAllocation() *big.Int {
	return a.balanceAllocat
}

// Balance returns the Shared account's balance, or nil if not yet synced.
func (a *NeuronAccount) Balance() *big.Int {
	return a.balance
}

// FR-018: Ledger attachment via derived EVM address
// LedgerAttachment returns the ledger attachment, or nil if not attached.
func (a *NeuronAccount) LedgerAttachment() *LedgerAttachment {
	a.mu.Lock()
	defer a.mu.Unlock()
	return a.ledgerAttach
}

// FR-022: Child registry binding (identifier + external id)
// RegistryBinding returns the registry binding for Child accounts, nil otherwise.
func (a *NeuronAccount) RegistryBinding() *RegistryBinding {
	return a.registryBinding
}

// FR-026: Child fee payer for transaction sponsorship
// FeePayer returns the fee payer ledger account ID for Child accounts, nil otherwise.
func (a *NeuronAccount) FeePayer() *LedgerAccountId {
	return a.feePayer
}

// P2PHost returns the p2p host PeerID for Child accounts (= PeerID), nil otherwise.
func (a *NeuronAccount) P2PHost() *keylib.PeerID {
	return a.p2pHost
}

// SetLedgerAttachment sets or replaces the ledger attachment on this account.
func (a *NeuronAccount) SetLedgerAttachment(la LedgerAttachment) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.ledgerAttach = &la
}

File: impl/typescript/src/account/account.ts
/**
 * NeuronAccount -- discriminated union type for agent identity.
 *
 * Spec reference: 001 spec.md
 *   - FR-001: Parent account with DID and single NeuronPublicKey.
 *   - FR-002: Child account with parent reference and own NeuronPublicKey.
 *   - FR-007a: Shared account with MultisigKey and threshold scheme.
 *   - FR-008: Derived fields evmAddress() and peerId() delegate to publicKey.
 *   - FR-013: AccountType discriminator (Parent=1, Child=2, Shared=3).
 *   - FR-016: Balance fields are optional (undefined until synced).
 *   - FR-018: LedgerAttachment is optional on all account types.
 *   - FR-022: All fields are readonly. Immutable after construction.
 *   - FR-026: FeePayer is optional on Child accounts.
 *
 * The NeuronAccount type is a discriminated union on accountType.
 * Type guards (isParent, isChild, isShared) narrow the union to
 * the specific variant for type-safe field access.
 */

import type { NeuronPublicKey } from '../keylib/public-key.js';
import type { EVMAddress } from '../keylib/evm-address.js';
import type { PeerID } from '../keylib/peer-id.js';
import type { MultisigKey } from '../keylib/multisig-key.js';
import type {
  NeuronDID,
  LedgerAttachment,
  RegistryBinding,
  LedgerAccountId,
} from './types.js';
import { AccountType } from './types.js';
import { invalidAccountType } from './errors.js';

/**
 * Base fields shared by all account types.
 *
 * FR-013: accountType is the discriminator.
 * FR-020: currencySymbol is required on all accounts.
 * FR-018: ledgerAttachment is optional on all accounts.
 */
export interface BaseAccount {
  /** FR-013: Account type discriminator. */
  readonly accountType: AccountType;
  /** FR-020: Currency symbol (e.g., "HBAR", "ETH"). */
  readonly currencySymbol: string;
  /** FR-018: Optional ledger attachment. */
  readonly ledgerAttachment?: LedgerAttachment | undefined;
}

/**
 * Parent account -- root identity with DID and credit balance.
 *
 * FR-001: Has a single NeuronPublicKey and a DID.
 * FR-006: Validation rules V-PARENT-01..05 apply.
 * FR-016: creditBalance is optional (undefined until synced from ledger).
 */
export interface ParentAccount extends BaseAccount {
  readonly accountType: AccountType.Parent;
  /** FR-001: Single secp256k1 public key. */
  readonly publicKey: NeuronPublicKey;
  /** FR-012: DID:key identifier derived from publicKey. */
  readonly did: NeuronDID;
  /** FR-016: Credit balance. Undefined until synced from ledger. */
  readonly creditBalance?: bigint | undefined;
}

/**
 * Child account -- agent identity with parent reference.
 *
 * FR-002: Has own NeuronPublicKey and a reference to Parent's publicKey.
 * FR-007: Validation rules V-CHILD-01..05 apply.
 * FR-016: balanceAllocation is optional (undefined until synced).
 * FR-022: registryBinding is mandatory for completeness.
 * FR-026: feePayer is optional.
 */
export interface ChildAccount extends BaseAccount {
  readonly accountType: AccountType.Child;
  /** FR-002: Single secp256k1 public key for this Child agent. */
  readonly publicKey: NeuronPublicKey;
  /** FR-007: Reference to the Parent's public key. */
  readonly parentPubKey: NeuronPublicKey;
  /** FR-016: Balance allocation from Parent. Undefined until synced. */
  readonly balanceAllocation?: bigint | undefined;
  /** FR-022: Registry binding. Mandatory for completeness (FR-011a). */
  readonly registryBinding?: RegistryBinding | undefined;
  /** FR-026: Fee payer account ID. Optional. */
  readonly feePayer?: LedgerAccountId | undefined;
}

/**
 * Shared account -- multi-signature account with threshold scheme.
 *
 * FR-007a: Has MultisigKey, no DID, no publicKey, no parent reference.
 * FR-021: balance is optional (undefined until synced).
 */
export interface SharedAccount extends BaseAccount {
  readonly accountType: AccountType.Shared;
  /** FR-007a: Multi-signature key configuration (m-of-n). */
  readonly multisigKey: MultisigKey;
  /** FR-021: Shared balance. Undefined until synced. */
  readonly balance?: bigint | undefined;
}

/**
 * Discriminated union of all account types.
 *
 * FR-013: Discriminated on accountType field.
 * Use type guards (isParent, isChild, isShared) to narrow.
 */
export type NeuronAccount = ParentAccount | ChildAccount | SharedAccount;

/**
 * Type guard: narrow NeuronAccount to ParentAccount.
 *
 * @param account - Account to check
 * @returns true if account is a Parent account
 */
export function isParent(account: NeuronAccount): account is ParentAccount {
  return account.accountType === AccountType.Parent;
}

/**
 * Type guard: narrow NeuronAccount to ChildAccount.
 *
 * @param account - Account to check
 * @returns true if account is a Child account
 */
export function isChild(account: NeuronAccount): account is ChildAccount {
  return account.accountType === AccountType.Child;
}

/**
 * Type guard: narrow NeuronAccount to SharedAccount.
 *
 * @param account - Account to check
 * @returns true if account is a Shared account
 */
export function isShared(account: NeuronAccount): account is SharedAccount {
  return account.accountType === AccountType.Shared;
}

/**
 * Derive the EVM address from an account's public key.
 *
 * FR-008: Delegates to publicKey.evmAddress() for Parent and Child accounts.
 * Shared accounts do not have a single public key and cannot derive an EVM address.
 *
 * @param account - NeuronAccount to derive EVM address from
 * @returns EVMAddress derived from the account's public key
 * @throws AccountError NEURON-ACCT-001 if account is Shared
 */
export function evmAddress(account: NeuronAccount): EVMAddress {
  if (isParent(account) || isChild(account)) {
    return account.publicKey.evmAddress();
  }
  throw invalidAccountType('Shared accounts do not have a single public key for EVM address derivation');
}

/**
 * Derive the libp2p PeerID from an account's public key.
 *
 * FR-008: Delegates to publicKey.peerId() for Parent and Child accounts.
 * Shared accounts do not have a single public key and cannot derive a PeerID.
 *
 * @param account - NeuronAccount to derive PeerID from
 * @returns PeerID derived from the account's public key
 * @throws AccountError NEURON-ACCT-001 if account is Shared
 */
export function peerId(account: NeuronAccount): PeerID {
  if (isParent(account) || isChild(account)) {
    return account.publicKey.peerId();
  }
  throw invalidAccountType('Shared accounts do not have a single public key for PeerID derivation');
}
